enhancements/cluster-constraints-properties.md

```diff
@@ -0,0 +1,172 @@
+- Provide mechanism to have separate namespaced and cluster-scoped properties/constraints
+
+## Proposal
+
+A new constraint API provides an effective away for cluster administrators to create custom resources (CR) to contain cluster information such as cluster constraints and properties.
+
+OLM resolver will retrieve the cluster constraints and properties from the CR(s) and the information can be used during resolution.
+
+The cluster constraints and properties are presented as a virtual cluster-scoped operator that is installed in the namespace. For all incoming operators that are intended to run the namespace, they must satisfy all constraints that are associated with this cluster-scoped operator. At the same time, the installable operators can depend on the properties that are provided by this virtual operator.
+
+### User Stories
+
+#### Cluster constraints
+
+As a cluster admin, I would like to specify constraints that are specified to the cluster and they that will be applied to all operators. As a result, only operators that satisfy these constraints to be installable in the cluster.
+
+#### Cluster properties
+
+As a cluster admin, I would like to provide cluster properties that are relevant to cluster availability and capabilities so that operator authors can depend on those properties if needed to ensure their operators are installable and functioning properly in the cluster.
+
+#### Bundle properties and constraints
+
+As an operator author, I would like to include constraints that rely on cluster properties so my operators in my bundles so that they can only be installed on suitable clusters.
+
+I also would like to provide essential properties that are used in cluster constraints in my bundle so that they can be screened at runtime if the clusters have certain requirements that may impact my operators' operation.
+
+### Design Details
+
+#### Cluster information API
+
+The cluster constraint API is constructed using CustomResourceDefinition (CRD):
+
+```yaml=
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: constraints.operators.coreos.com
+spec:
+  group: operators.coreos.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                constraints:
+                  type: string
+                properties:
+                  type: string
+  scope: Cluster
+  names:
+    plural: constraints
+    singular: constraint
+    kind: constraint
+    shortNames:
+    - constraint
+```
+
+The CRD is cluster-scoped as the CR(s) are representing cluster information. The cluster administrators have RBAC permission to create CustomResource (CR) that will contain cluster constraints and properties. For example:
+
+```yaml=
+apiVersion: operators.coreos.com/v2
+kind: Constraint
```

Comment:
Can we make this `PropertySet` or something? Isn't a `constraint` just a specifial form of property?